waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -1704,6 +1738,170 @@
+  "Query K8s for all namespace-scoped events with matching object names. Returns a
+   channel containing the response or an exception."
+  [{:keys [api-server-url] :as scheduler} {:keys [api-request-fn]} {:keys [namespace k8s-object-name]}]
+  (log/info "fetching events from k8s for" namespace k8s-object-name)
+  (let [events-url (str api-server-url "/api/v1/namespaces/" namespace "/events?fieldSelector=involvedObject.name%3D" k8s-object-name)
+        events-chan (api-request-fn events-url scheduler)]
+    events-chan))
+
+(defn get-cached-events
+  "Load events from cache or query K8s and store the channel that will contain the response."
+  [k8s-object-key->event-cache k8s-object-key k8s-query-fn scheduler options]
+  (cu/cache-get-or-load k8s-object-key->event-cache k8s-object-key
+                        (fn load-event-from-k8s [] (k8s-query-fn scheduler options k8s-object-key))))
+
+(defn service-unhealthy?
+  "Return true if the given service appears unhealthy"
+  [{:keys [instances k8s/replicaset-creation-timestamp task-count task-stats]}]
+  (let [creation-timestamp (du/str-to-date-safe replicaset-creation-timestamp)
+        healthy-count (:healthy task-stats 0)
+        thirty-seconds-ago (t/ago (t/seconds 30))]
+    (and (or (nil? creation-timestamp)
+             (t/before? creation-timestamp thirty-seconds-ago))
+         (or (< task-count instances)
+             (< healthy-count task-count)))))
+
+(defn pod-unhealthy?
+  "Return true if the given pod was created more than 30 seconds ago and it appears unhealthy"
+  [{:keys [metadata status]}]
+  (let [creation-timestamp (timestamp-str->datetime (:creationTimestamp metadata))
+        thirty-seconds-ago (t/ago (t/seconds 30))
+        {:keys [conditions]} status
+        condition-statuses (map (comp #(Boolean/valueOf %) :status) conditions)]
+    (and (or (nil? creation-timestamp)
+             (t/before? creation-timestamp thirty-seconds-ago))
+         (or (nil? condition-statuses)
+             (empty? condition-statuses)
+             (not-every? true? condition-statuses)))))
+
+(defn- get-service-key
+  "Return a k8s-object key using the namespace and app name of the given service."
+  [service]
+  (let [namespace (:k8s/namespace service)
+        name (:k8s/app-name service)]
+    {:namespace namespace :k8s-object-name name}))
+
+(defn- get-pod-key
+  "Return a k8s-object key using the namespace and name from the given pod's metadata."
+  [{:keys [metadata]}]
+  (let [namespace (:namespace metadata)
+        name (:name metadata)]
+    {:namespace namespace :k8s-object-name name}))
+
+(defn- get-events-for-k8s-object
+  "Get K8s Events for the given object. If Events are cached, they will be returned,
+   otherwise a channel will be returned. This channel will contain the fetched events
+   or an exception."
+  [k8s-object get-object-key load-events]
+  (let [k8s-object-key (get-object-key k8s-object)
+        events-or-chan (load-events k8s-object-key)]
```

Comment:
Can we consolidate to a single type being returned by `load-events`? We can wrap values inside promise chans if required.